dex/encode/passbytes.go

```diff
@@ -0,0 +1,266 @@
+// UTF8-encoded byte slice.
+func (pb PassBytes) MarshalJSON() ([]byte, error) {
+	// pb may have been created by calling PassBytes("some invalid string").
+	// Returning the quoted copy of pb may lead to errors later when trying to
+	// unmarshall. Sanity check that pb is a valid UTF8-encoded byte slice.
+	if !isUTF8Encoded(pb) {
+		return nil, fmt.Errorf("invalid PassBytes data")
+	}
+	data := make([]byte, len(pb)+2)
+	data[0], data[len(data)-1] = '"', '"'
+	copy(data[1:], pb)
+	return data, nil
+}
+
+// UnmarshalJSON satisfies the json.Unmarshaler interface, parses JSON-encoded
+// data into UTF8-encoded bytes and stores the result in the `PassBytes` pointer.
+func (pb *PassBytes) UnmarshalJSON(rawBytes []byte) error {
+	utf8EncodedBytes, err := parseJSONEncodedDataAsUTF8Bytes(rawBytes)
+	ClearBytes(rawBytes)
+	if err != nil {
+		return fmt.Errorf("cannot unmarshal password: %v", err)
+	}
+	*pb = utf8EncodedBytes
+	return nil
+}
+
+// Clear zeroes the slice.
+func (pb PassBytes) Clear() {
+	ClearBytes(pb)
+}
+
+func isUTF8Encoded(data []byte) bool {
+	if len(data) == 0 {
+		return true // represents an empty string
+	}
+	if data[0] == '"' {
+		// should not be quoted, quotes must be escaped!
+		return false
+	}
+
+	readIndex := 0
+	for readIndex < len(data) {
+		byteAtPos := data[readIndex]
+		switch {
+		case byteAtPos == escapeSequence:
+			// Escape sequence hit, expect a valid escape char from next byte
+			// (or byte sequence).
+			if readIndex+1 >= len(data) {
+				return false
+			}
+			nextByte := data[readIndex+1]
+			if nextByte == unicodePrefix {
+				// expect a unicode char in the form \uXXXX
+				// or a surrogate pair in the form \uXXXX\uYYYY
+				_, bytesRead := unicodeSequenceToCharacter(data[readIndex:])
+				if bytesRead <= 0 {
+					return false
+				}
+				readIndex += bytesRead
+			} else {
+				// some other escaped character?
+				_, ok := parseEscapedCharacter(nextByte)
+				if !ok {
+					return false
+				}
+				readIndex += 2
+			}
+
+		case byteAtPos == '"', byteAtPos < ' ':
+			// invalid char
+			return false
+
+		default:
+			// Attempt to decode char as UTF8, may get utf8.RuneError.
+			_, bytesRead := utf8.DecodeRune(data[readIndex:])
+			if bytesRead <= 0 {
+				return false
+			}
+			readIndex += bytesRead
+		}
+	}
+
+	// all bytes check out
+	return true
+}
+
+// parseJSONEncodedDataAsUTF8Bytes parses the provided JSON-encoded data into a
+// UTF8-encoded byte slice.
+// Returns an error if any of the following conditions is hit:
+// - `data` is not a valid JSON encoding
+// - `data` is not quoted
+// - `data` contains a byte or byte sequence that cannot be parsed into a
+//    UTF8-encoded byte or byte sequence.
+// Inspired by encoding/json.(*decodeState).unquoteBytes.
+func parseJSONEncodedDataAsUTF8Bytes(data []byte) ([]byte, error) {
+	if !json.Valid(data) {
+		return nil, fmt.Errorf("data is not json-encoded")
+	}
+	if data[0] != '"' || len(data) < 2 || data[len(data)-1] != '"' {
```

Comment:
```suggestion
	if len(data) < 2 || data[0] != '"' || data[len(data)-1] != '"' {
```

Just trying out the github suggestions here.  This seems to be it for this PR. Nicely done!

Reply:
Thanks @chappjc! Suggestions are pretty nice, I could just directly commit your suggestion but I want to address buck's other comment on this same code section, so I'll commit from my local branch.

Reply:
Cool.  I'm actually not sure what the git mechanics of applying a suggestion are, but I imagine it's something messy so best to do it manually.